Can't make/gmake on FreeBSD 12.1

Platform: FreeBSD 12.1 64-bit (running inside Vmware Workstation Player 15) - 4 GB RAM
libxcrypt Release: 4.4.16

1. The Installation file says to use 'make' .. but when I try ./configure MAKE="make" it give me errors. 
a. bash ./autogen.sh output : http://bsd.to/cmc4
b. ./configure or ./configure MAKE="make" .. it gives me following error/output: https://bsd.to/nxDe

2. When I use gmake (4.3) instead using ./configure MAKE="gmake" 
b. running ./configure completes and give this output: https://bsd.to/sMXe
c. but breaks when I do gmake: http://bsd.to/SZzH

Hi, thanks for your interest in using libxcrypt on FreeBSD.  We would be happy to consider patches to improve binary compatibility with the system-provided libcrypt on this platform.

> ```
> config.status: error: in `/usr/home/user123/libxcrypt/libxcrypt-4.4.16':
> config.status: error: Something went wrong bootstrapping makefile fragments
>  for automatic dependency tracking. If GNU make was not used, consider
>  re-running the configure script with MAKE="gmake" (or whatever is
>  necessary). You can also try re-running configure with the
>  '--disable-dependency-tracking' option to at least be able to build
>  the package (albeit without support for automatic dependency tracking).
> See `config.log' for more details
> ```

I'm able to reproduce this on a FreeBSD 11 host, using autoconf 2.69 and automake 16.1.  This is a limitation in Automake; it does not support BSD make adequately.  There's not much I can do about it other than adding a note to the README about using either `--disable-dependency-tracking` or GNU make. You might consider reporting the problem to the Automake maintainers (see https://www.gnu.org/software/automake/manual/html_node/Reporting-Bugs.html ).

> ```
> In file included from lib/randombytes.c:35:
> /usr/include/unistd.h:489:8: error: redefinition of 'struct crypt_data'
>  489 | struct crypt_data {
>      |        ^~~~~~~~~~
> In file included from lib/crypt-port.h:397,
> from lib/randombytes.c:17:
> ./crypt.h:73:8: note: originally defined here
> 73 | struct crypt_data
>    |        ^~~~~~~~~~
> ```

This, on the other hand, I cannot reproduce on FreeBSD 11.  I don't have access to a FreeBSD 12 host, but I understand what's wrong anyway.  It's going to be a little awkward to fix and I don't have time to work on it today, but I may ask you to test a pull request sometime later this week.

Hi Besser82 

I just tested it with the following:
`git clone --single-branch --branch topic/besser82/issue_110 https://github.com/besser82/libxcrypt.git
./autogen.sh
./configure MAKE="gmake" --disable-dependency-tracking
gmake
`

didn't work, got a new error: 
`/usr/local/bin/ld: .libs/libcrypt.so.2.0.0: version node not found for symbol crypt_checksalt@@XCRYPT_4.3
/usr/local/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
gmake[1]: *** [Makefile:1449: libcrypt.la] Error 1
gmake[1]: Leaving directory '/usr/home/user123/libxcrypt/libxcrypt'
gmake: *** [Makefile:1272: all] Error 2`

--- a/NEWS
+++ b/NEWS
@@ -4,6 +4,7 @@ Please send bug reports, questions and suggestions to
 <https://github.com/besser82/libxcrypt/issues>.
 
 Version 4.4.18
+* Fix compilation errors on (Free)BSD (issue #110).
 
 Version 4.4.17
 * Fix compilation error in 'alignas (type)' with older versions

--- a/lib/crypt-port.h
+++ b/lib/crypt-port.h
@@ -50,6 +50,23 @@
 #include <sys/param.h>
 #endif
 
+/* The prototypes of the crypt() function and struct crypt_data possibly
+   defined in <unistd.h> need to be renamed as they might be incompatible
+   with our declaration.
+   Defining these macros *AFTER* <crypt-symbol-vers.h>, which also defines
+   the same macros for symbol-versioning purposes, was included, would lead
+   to a clashing redefinition of these macros, and thus cause our internal
+   symbol-versioning would not work properly anymore.  */
+#ifdef HAVE_UNISTD_H
+#define crypt unistd_crypt_is_incompatible
+#define crypt_r unistd_crypt_r_is_incompatible
+#define crypt_data unistd_crypt_data_is_incompatible
+#include <unistd.h>
+#undef crypt
+#undef crypt_r
+#undef crypt_data
+#endif
+
 #ifndef HAVE_SYS_CDEFS_THROW
 #define __THROW /* nothing */
 #endif

--- a/lib/randombytes.c
+++ b/lib/randombytes.c
@@ -31,9 +31,6 @@
 #ifdef HAVE_SYS_STAT_H
 #include <sys/stat.h>
 #endif
-#ifdef HAVE_UNISTD_H
-#include <unistd.h>
-#endif
 
 /* If we have O_CLOEXEC, we use it, but if we don't, we don't worry
    about it.  */

--- a/test/badsalt.c
+++ b/test/badsalt.c
@@ -22,7 +22,6 @@
 #include <stdio.h>
 #include <string.h>
 #include <sys/mman.h>
-#include <unistd.h>
 
 static const char phrase[] = "values of β will give rise to dom!";
 

--- a/test/crypt-badargs.c
+++ b/test/crypt-badargs.c
@@ -78,7 +78,7 @@ static const char *settings[] =
 };
 
 /* In some of the tests below, a segmentation fault is the expected result.  */
-static jmp_buf env;
+static sigjmp_buf env;
 static void
 segv_handler (int sig)
 {

--- a/test/crypt-badargs.c
+++ b/test/crypt-badargs.c
@@ -15,7 +15,6 @@
 #include <string.h>
 #include <signal.h>
 #include <sys/mman.h>
-#include <unistd.h>
 
 /* The behavior tested below should be consistent for all hashing
    methods.  */

--- a/test/getrandom-fallbacks.c
+++ b/test/getrandom-fallbacks.c
@@ -23,9 +23,6 @@
 #ifdef HAVE_SYS_STAT_H
 #include <sys/stat.h>
 #endif
-#ifdef HAVE_UNISTD_H
-#include <unistd.h>
-#endif
 
 /* If arc4random_buf is available, all of the fallback logic is compiled
    out and this test is unnecessary.  If ld --wrap is not available this

--- a/test/getrandom-interface.c
+++ b/test/getrandom-interface.c
@@ -125,7 +125,7 @@ test_fault (char *page, size_t pagesize)
 }
 
 /* In one of the tests above, a segmentation fault is the expected result.  */
-static jmp_buf env;
+static sigjmp_buf env;
 static void
 segv_handler (int sig)
 {

--- a/test/getrandom-interface.c
+++ b/test/getrandom-interface.c
@@ -15,7 +15,6 @@
 #include <string.h>
 #include <signal.h>
 #include <sys/mman.h>
-#include <unistd.h>
 
 static bool error_occurred;
 

